# Using SEGUL

## Command Structure

The app command structure is similar to git, gh-cli, or any other app that use subcommands. The app file name will be `segul` for Linux/MacOS/WSL and `segul.exe` for Windows.

```Bash
[THE-PROGRAM-FILENAME] <SUBCOMMAND> [OPTIONS] <VALUES> <A-FLAG-IF-APPLICABLE>
```

To check for available subcommand:

```Bash
segul --help
```

To check for available options and flags for each sub-command:

```Bash
segul <SUBCOMMAND> --help
```

Across the app functions, most generic arguments are also available in short format to save time typing them. For example, below we use short arguments to concat alignments in a directory named `nexus-alignments`:

```Bash
segul concat -d nexus-alignments -f nexus
```

## Input Options

The app has two input options. The standard input `--input` or `-i` in short format and `--dir` or `-d` in short format. If your input files are all in a single directory, you should use the `--dir` or `-d` option and specify the file format:

```Bash
segul <SUBCOMMAND> -d alignment_dir -f nexus
```

When dealing with a single file, more complex folder structure, or unusual file extensions, use the `--input` or `-i` option.

For a single file:

```Bash
segul <SUBCOMMAND> -i alignment-dir/alignment_file.fasta
```

Multiple file in a directory using wildcard:

```Bash
segul <SUBCOMMAND> -i alignment-dir/*.fasta
```

Multiple files in multiple directories:

```Bash
segul <SUBCOMMAND> -i alignment-dir1/*.fasta alignment-dir2/*.fasta
```

For unusual file extensions or if the app failed to detect the file format, specify the input format:

```Bash
segul <SUBCOMMAND> -i alignment-dir/*.aln -f fasta
```

Both of the input options are available in all subcommands.

## Datatype

The app support both DNA and amino acid sequences. It will check whether the sequences contain only valid IUPAC characters of the datatype. By default, it sets to DNA sequences. Use the option `--datatype aa` if your input is amino acid sequences. For example:

```Bash
segul convert -d /alignments -f nexus --datatype aa
```

## Output

Most functions will save into their default directory. For example, the concat function will default to create `SEGUL-concat` directory and will save its output files into the directory. To specify the output directory, use the `--output` or `-o` option. For example:

```Bash
segul convert -d /alignments -f nexus -o alignments_concat
```

By default, the app avoids over-writing files with similar names. The app will check if a such file or directory exists and will ask if you would like to remove it. The app will exit if you decide to not remove it. You can pass `--overwrite` flag to remove existing output files or directories without asking first.

## Computing sequence summary statistics

`segul` generate summary statistics for all alignment, each locus, and each taxon. To generate the summary statistics in for alignment files in a directory (usign `--dir` or `-d` option):

```Bash
segul summary --dir [a-path-to-a-directory] --input-format [sequence-format-keyword]
```

Using `--input` (or `-i` in short version) option:

```Bash
segul summary --input [a-path/wilcard-to-files]
```

`segul` can also generate summary statistics for each alignment by passing `--per-locus` flag:

```Bash
segul summary --dir [a-path-to-a-directory] --input-format [sequence-format-keyword] --per-locus
```

## Concatenating alignments

To concat all alignments in a directory using `--dir` (or `-d` in short version):

```Bash
segul concat --dir [a-path-to-a-directory] --input-format [sequence-format-keyword]
```

Using `--input` (or `-i` in short version) option:

```Bash
segul concat --input [path/wildcard-to-your-files]
```

## Converting sequence formats

Segul can convert a single sequence file or multiple sequence files in a directory. To input the file, use `--dir` or `-d` option:

```Bash
segul convert --dir [path-to-your-repository] --input-format [sequence-format-keyword] --output-format [sequence-format-keyword]
```

Using `--input` (or `-i` in short version) option:

```Bash
segul convert --input [path/wildcard-to-your-files] --output-format [sequence-format-keyword]
```

## Extracting sequences in alignments

You can also extract sequences from a collection of alignments. It can be done by supplying a list of IDs directly using the command line or in a text file. `segul` will look for the exact match of the IDs and is case sensitive. You can also use regular expression for more flexiple option.

To extract sequences by inputing the IDs in the command line:

```bash
segul extract --dir [path-to-alignment-directory] --input-format [sequence-format-keyword] --id [id_1] [id_2] [id_3]
```

Using `--input` (or `-i` in short version) option:

```bash
segul extract --input [a-path/wilcard-to-files] --id [id_1] [id_2] [id_3]
```

You can specify as many id as you would like. However, for a long list of IDs, it may be more convenient to input a text file. In the file, the content should be only the list of IDs, one ID each line:

```bash
sequence_1
sequence_2
sequence_3
sequence_4
```

The the command will be:

```bash
segul extract --dir [path-to-alignment-directory] --input-format [sequence-format-keyword] --file [path-to-text-file]
```

For using regular expression:

```bash
segul extract -d gblock_trimmed_80p/ -f nexus --re="regex-syntax"
```

`segul` uses the rust [regex](https://docs.rs/regex/1.5.4/regex/) to parse regular expression. The syntax is similar to Perl regular expression.

## Filtering alignments

Segul provide multiple filtering parameters (see below).

Using a directory `--dir` (or `-d` in short version):

```Bash
segul filter --dir [a-path-to-a-directory] --input-format [sequence-format-keyword] <parameters>
```

Using `--input` (or `-i` in short version) option:

```Bash
segul filter --input [a-path/wilcard-to-files] <parameters>
```

For example, to filter based on taxon completeness:

```Bash
segul filter --dir [a-path-to-a-directory] --input-format [sequence-format-keyword] --percent [percentages-of-minimal-taxa]
```

Other available parameters are multiple minimal taxon completeness `--npercent`, alignment length `--len`, numbers of minimal parsimony informative sites `--pinf`, and percent of minimal parsimony informative sites `--percent-inf`.

By default, the app will copy files that are match with the parameter to a new folder. If you would like to concat the results instead, you can specify it by passing `--concat` flags. All the options available for the concat function [above](#concatenating-alignments) also available for concatenating filtered alignments.

## Getting sample IDs from a collection of alignments

You have multiple alignments and want to know what are samples you have in all of those alignment. You can easily do it using `segul`. The app can find all the unique IDs across thousands of alignments within seconds.

The command using a `--dir` (or `-d` in short version):

```Bash
segul id --dir [a-path-to-a-directory] --input-format [sequence-format-keyword]
```

Using `--input` (or `-i` in short version) option:

```Bash
segul id --input [a-path/wilcard-to-files]
```

It will generate a text file that contains all the unique IDs across your alignments.

## Map sample distribution in a collection of alignments

If you would like to know how the samples distributed across your alignments, you only need to add the `--map` flag when searching for unique IDs. It will generate both the unique IDs (in a text file) and the sample distribution in boolean format (save to a csv file).

Using a `--dir` (or `-d` in short version):

```Bash
segul id --dir [a-path-to-a-directory] --input-format [sequence-format-keyword] --map
```

Using `--input` (or `-i` in short version) option:

```Bash
segul id --input [a-path/wilcard-to-files] --map
```

## Batch removing sequence based on user-defined IDs

Based on a list of IDs, you can remove sequences in a collection of alignments.

```Bash
segul remove --dir [alignment-dir] -f [sequence-format-keyword] --id [list-of-id]
```

Using regular expression:

```Bash
segul remove --dir [alignment-dir] -f [sequence-format-keyword] --re=["regex"]
```

If you remove more than a half of the sequence, `segul`'s extract feature is more efficient.

## Batch renaming sequence IDs

To rename sequence IDs in multiple alignments, you need to input the sequence IDs in tsv or csv format with header. For example:

| Original_names        | New_names                |
| --------------------- | ------------------------ |
| Genus_species1_random | Genus_species1_voucherID |
| Genus_species2_random | Genus_species2_voucherID |

To simplify this process, you can generate unique IDs for all of your alignments using the `id` sub-command.

```Bash
segul id --dir [alignment-dir] -f [sequence-format-keyword]
```

Copy the IDs to a spreadsheet application, such as Microsoft Excel and then write new names and the header names as above. Save the file as csv or tsv. The program will infer the file format based on the file extension. Use it as an `--names` or `-n` input:

```Bash
segul rename --dir [alignment-dir] -f [sequence-format-keyword] --replace [file-path-to-IDs]
```

Using `--input` (or `-i` in short version) option:

```Bash
segul rename --input [a-path/wilcard-to-files] --replace [file-path-to-IDs]
```

Example:

```Bash
segul rename -d uce-loci/ -f nexus --replace new_names.csv
```

You can also remove a part of the sequence name using a string input. Command below will remove each occurrence of `LOCUS` in the sequence IDs.

```Bash
segul rename -d -d uce-loci/ -f nexus --remove="LOCUS"
```

Using regular expression:

```Bash
segul rename -d -d uce-loci/ -f nexus --remove-re="(?i)^LOCUS-\d{3}"
```

The regex above match the first occurence of `locus-` (case insensitive) followed by 3 digits. For multiple occurrences in a single sequence name, use the option `--remove-re-all`. If your sequence names as below:

```Text
species_epithet_locus-100
species_epithet_locus-102
species_epithet_locus-103
```

They will change to:

```Text
species_epithet
species_epithet
species_epithet
```

You can also replace a part of the sequence name with other string:

```Bash
segul rename -d -d uce-loci/ -f nexus --replace-from="GENE" --replace-to="LOCUS
```

Using regular expression:

```Bash
segul rename -d -d uce-loci/ -f nexus --replace-from-re="(?i)GENE" --replace-to="LOCUS
```

You can also change the output format by using `--output-format` or `-F` option.

## Translating DNA sequences

`segul` translate DNA sequence based on [NCBI Genetic Code Tables](https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi#top).

To translate dna alignment to amino acid using `--dir` (or `-d` in short version):

```Bash
segul translate --dir [path-to-alignment-files] -f [sequence-format-keyword]
```

Using `--input` (or `-i` in short version) option:

```Bash
segul translate --input [a-path/wilcard-to-files]
```

By default, the app will use the NCBI standard code table (Table 1). To set the translation table, use the `--table` option. For example, to translate dna sequences using NCBI Table 2 (vertebrate MtDNA):

```Bash
segul translate -d loci/ -f fasta --table 2
```

You can also set the reading frame using the `--rf` option:

```Bash
segul translate -d loci/ -f fasta --table 2 --rf 2
```

To show all the table options, use the `--show-tables` flag:

```Bash
segul translate --show-tables
```

## Log Files

Most information that is printed to the terminal is written to the log file (named `segul.log`). It is written to the current working directoy. Unlike the terminal output that we try to keep it clean and only show the most important information, the log file will also contain the dates, times, and the log level status. Each time you run the app, if the log file exists in the same directory, the app will append the log output to the same log file. Rename this file or move it to a different directory if you would like to keep a different log file for each task.
