## Concatenating alignments

`segul` provides an easy way to concat multiple alignments and generate the partition setting at the same time.

```Bash
segul concat <input-argument> [input-path] --input-format [sequence-format]
```

For example, we will concat all the alignments in this folder:

```Bash
alignments/
â”œâ”€â”€ locus_1.nexus
â”œâ”€â”€ locus_2.nexus
â””â”€â”€ locus_3.nexus
```

We can do it in two ways. First, using `--dir` argument to input alignment files:

```Bash
segul concat --dir alignments/ --input-format nexus
```

Second, we can use the `--input` argument to input the alignment files. We will rely on wildcard and the OS to find all the alignment files. We do not need to specify the input format for `.nexus` file extension:

```Bash
segul concat --input alignments/*.nexus
```

`segul` will generate two files saved in `SEGUL-concat` directory, consisting of the concatenated alignments and the partition settings:

```Bash
SEGUL-concat/
â”œâ”€â”€ SEGUL-Concat.nex
â””â”€â”€ SEGUL-Concat_partition.nex
```

To specify the name of the output directory, use the `--output` or `-o` option. Below, we will name our output directory `aln-concat`.

```Bash
segul concat --input alignments/*.nexus --output aln-concat
```

To specify the prefix of the file names use the `--prefix` option. Below, our output filenames will start with concat:

```Bash
segul concat --input alignments/*.nexus --output aln-concat --prefix concat
```

The resulting output directory will contain files as below:

```Bash
aln-concat/
â”œâ”€â”€ concat.nex
â””â”€â”€ concat_partition.nex
```

By default, the partition format is in nexus:

```Text
#nexus
begin sets;
charset 'locus-1' = 1-666;
charset 'locus-2' = 667-1473;
charset 'locus-3' = 1474-2000;
end;
```

You can specify the partition format using `--part` or `-p` option.

For example, to use raxml format:

```Bash
segul concat --input alignments/*.nexus --output concat --prefix concat --part raxml
```

The resulting partioned will be formatted in raxml style:

```Text
DNA, locus_1 = 1-666
DNA, locus_2 = 667-1473
DNA, locus_3 = 1474-2000
```

If the input is amino acid sequence, the partion will not contain the datatype:

The resulting partioned will be formatted in raxml style:

```Text
locus_1 = 1-666
locus_2 = 667-1473
locus_3 = 1474-2000
```

You can also use `charset` format. In this format, the partition will be written at the end of the sequence and only available for the nexus output. This format is usually required for phylogenetic programs, such PAUP and BEAST. To use `charset` format:

```Bash
segul concat --input alignments/*.nexus --output concat --prefix concat --part charset
```

You can also write the partition to a codon model format by using the flag `--codon`. You may not need this option for genomic datasets. We reserve this function for Sanger datasets.

## Converting sequences to a different format

Segul can convert a single file, multiple file in the same directory, or multiple files in multiple directory.

### Converting a single file

To convert a single file, use the `--input` or `-i`.

```Bash
segul convert --input [a-path-to-file] --output-format [sequence-format]
```

`segul` should be able to infer the input file format based on the extension of your file. If your file extension is uncommon (for example .aln instead of .fas), specify the `--input-format`:

```Bash
segul convert --input [a-path-to-file] --input-format [sequence-format] --output-format [sequence-format]
```

For example, here we will convert a nexus file containing DNA sequence named `loci.nexus` to fasta format:

```Bash
segul convert --input alignments/loci.nexus -output-format fasta
```

You can ignore specifying the output format if you would like the output files in nexus format. For example, we will convert loci.fasta to nexus format:

```Bash
segul convert --input alignments/loci.fasta
```

Some `segul` arguments available in short format, notice the uppercase 'F' for the output format:

```Bash
segul convert -i alignments/loci.nexus -f nexus  -F fasta --datatype aa -o alignment-fasta
```

> Note that, when converting alignments, `segul` will only change the file extension and will maintain the original file names for the output files. This behavior is the same for a single or multiple file format conversion. This way it gives the app the flexibility to convert many files in a single command.

### Batch converting sequence files in a directory

We have to option to provide the input files to the progam. First, using the `--dir` or `-d` input. For a directory input, it is required to specify the `--input-format`:

```Bash
segul convert --dir [path-to-your-repository] --input-format [sequence-format] --output [your-output-dir-name] --output-format [sequence-format]
```

In short format:

```Bash
segul convert -d [path-to-your-repository] -f [sequence-format] -o [your-output-dir-name]
```

For example, suppose we want to convert all the nexus files in the directory below to fasta formats and name the output directory `alignments-fas`:

```Bash
alignments/
â”œâ”€â”€ locus_1.nexus
â”œâ”€â”€ locus_2.nexus
â””â”€â”€ locus_3.nexus
```

The command will be:

```Bash
segul convert -d alignments/ -f nexus -F fasta -o alignments-fas
```

We can also input wildcard (\*) using the `--input` or `-i` option to achieve the same results:

```Bash
segul convert -i alignments/*.nexus -f nexus -F fasta -o alignments-fas
```

The outputs will be:

```Bash
alignments-fas/
â”œâ”€â”€ locus_1.fas
â”œâ”€â”€ locus_2.fas
â””â”€â”€ locus_3.fas
```

### Batch converting multiple files in multiple directories

To convert multiple files in multiple directories, we will inputs multiple paths with wildcards and use the `--input` or `-i` option:

```Bash
segul convert -i [wildcard-1] [wildcard-2] [wildcard-3] -o [your-output-dir-name]
```

For example:

```Bash
segul convert -i alignments-1/*.nexus alignments-2/*.nexus alignments-3/*.nexus -F fasta -o alignments-fas
```

You can also convert multiple files in different formats. Because `segul` maintain the original file name for the output, the filename for each input file should be unique regardless of the extension. Otherwise, `segul` will failed to write the output.

For example:

```Bash
segul convert -i alignments-1/*.nexus alignments-2/*.fasta alignments-3/*.fasta -F phylip -o alignments-phylip
```

Depending on the size of your RAM, you can provide as many input paths as possible. The output files, however, will all be saved in the same directory.

### Converting amino acid sequences

By default, `segul` datatype is set for converting DNA sequence. If your file contains amino acid sequence, use the argument `--datatype aa`. For example:

```Bash
segul convert --input alignments/loci.nexus -output-format fasta --datatype aa
```

### Specifying the output directory

By default segul will write the result in a directory called `SEGUL-convert`. If you would like to specify the output directory, use the `--output` argument. For example, here we will specify the output directory to `alignment-fasta`.

```Bash
segul convert --input alignments/loci -output-format fasta --datatype aa --output alignment-fasta
```

### Sorting the output sequences

By default, `segul` maintain the original order of the sequences in the input file(s). Using the `--sort` flag, you can sort the sequences in alphabetical order based on their IDs:

```Bash
segul convert --input alignments/loci -output-format fasta --datatype aa --output alignment-fasta --sort
```

### Terminal output for file conversion

For sequence format conversion, `segul` terminal output consists of three parts. The first part is the input information. It contains information from your command input. The second part is the spinner. It will show you the state of the program as it is processing the file. The third part is the output information. It shows the information of the output directory relative to the current working directory, the output format, and the log file path. With the exception of the spinner (include the spinning emoji and its text), all the terminal output is written into the log file.

Below is an example of terminal output when you use the argument `--input` or `-i` to input the files:

```Text
=========================================================
SEGUL v0.11.1
An alignment tool for phylogenomics
---------------------------------------------------------
Input             : STDIN
File counts       : 1
Input format      : Auto
Data type         : DNA
Task              : Sequence format conversion

ðŸŒ˜ Finished converting sequence format!

Output
Output dir        : SEGUL-Convert
Output format     : Nexus Sequential
Log file          : segul.log

Execution time    : 9.7194ms
```

For `--dir` or `-d` input, notice the input file path is printed in the terminal:

```Text
=========================================================
SEGUL v0.11.1
An alignment tool for phylogenomics
---------------------------------------------------------
Input dir         : align-fas/
File counts       : 1
Input format      : Fasta
Data type         : DNA
Task              : Sequence format conversion

ðŸŒ˜ Finished converting sequence format!

Output
Output dir        : SEGUL-Convert
Output format     : Nexus Sequential
Log file          : segul.log

Execution time    : 13.5741ms
```

## Converting partition formats

`segul` can convert a single and multiple partition files in multiple folders. You can also use this function to extract partition embedded in NEXUS sequence files.

```Bash
segul partition --input <a-path/wildcard-to-partition> --input-part <input-partition-format> --output-part<output-partition-format>
```

If you output to RaXML partition and it is a DNA sequence, we recommend to use `--datatype dna` option. The partition format will include the datatype as below:

```Text
DNA, locus_1 = 1-100
DNA, locus_2 = 101-150
DNA, locus_3 = 151-200
```

By default, `segul` check the partition format:

1. First partition position starts with 1.
2. The next partition is next number after the end of the previous partition. If codon model, it will check the next locus applies the rule.

You can ignore format checking by passing the `--uncheck` flag.

For example, to extract nexus in-file partitions (called charset format in `segul`):

```Bash
segul partition --input concatenated_alignment.nex --input-part charset --output-part nexus
```

You can also use wildcard to convert multiple concatenated alignment partitions at once:

```Bash
segul partition --input ./*/concatenated_alignment.nex --input-part charset --output-part nexus
```

`segul` by default will save the resulting partitions in the parent directory of the input partition (or alignment if the partition is charset).

## Extracting sequences in alignments

`segul` can extract sequences in a collection of alignments based on the sequence ID. You can input the sequence ID in three ways:

1. Using the standard input (stdin) via terminal
2. By inputing a list of IDs in a text file
3. Using regular expression

By default, segul will write the results in a nexus format and use the name from the original input. More details [below](#specifying-output-directory-and-format).

### Extracting sequences using stdin

This method is useful if you only need to extract a few sequences across all your alignmetns. To input the list of ID, use `--id` option. It allow as many sequence ID as you can type in your terminal. For a long list, however, we recommend to input the ID in a text file.

```Bash
segul <input-option> [alignment-dir] --input-format [seq-format] --id [list-of-sequence-id]
```

For example, using the command below, we will extract `seq_1`, `seq_2`, and `seq_3` from our alignments stored in `alignments/` directory:

```Bash
segul --dir alignments/ --input-format nexus --id seq_1 seq_2 seq_3
```

### Extracting sequences using an input file

With a long list of IDs, it will be more convenient to input the list of the IDs instead. `segul` `--file` option is available for this task. You only need to list all the IDs in a text file with each ID in a new line as below:

```Text
seq_1
seq_2
seq_3
seq_4
seq_5
```

The command will be:

```Bash
segul extract --dir alignments/ --input-format nexus --file id.txt
```

For lots of ID, you can generate all the ID first using unique ID finder function on `segul` ([see above](#finding-unique-ids-in-alignments)). Open the text file using your text editor and remove the sequence IDs that you don't need. Use this file as an input for `segul extract`.

> NOTES: `segul` can parse input file written both on Unix-like OS (Linux, WSL, and MacOS) and Windows with carriage return.

### Extracting sequences using regular expression

This is the most flexible option to extract sequence from a collection of alignments. `segul` use the Rust [regex](https://docs.rs/regex/latest/regex/) library to parse regular expression syntax. The syntax is similar to Perl-style regex syntax (details [here](https://docs.rs/regex/latest/regex/#syntax)). You can use [this website](https://regex101.com/) to test your syntax.

To use regular expression, use the option `--re=` and put the syntax in either a single or double quatation. For example, our sequence id is named this way `genus_species_voucherNo`. We want to extract the sequences from the same genus: `Mus`. We could write a regular expression that search for sequence ID that starts with `Mus` as below:

```Bash
segul extract --dir alignments/ --input-format nexus --re="^Mus"
```

We could also search it with case insentive match:

```Bash
segul extract --dir alignments/ --input-format nexus --re="^(?i)mus"
```

All the matched sequences found in each alignment will be written to the output files. The name of the alignment will be maintain as the original.

### Specifying output directory and format

By default, the output files will be written into `SEGUL-Extract` directory. To change the directory name, use the `--ouput` or `-o` option. For the output files, the matched sequences will be written in nexus format using the alignment original name. You can change the output format using `--output-format` (or `-F` in short version) option. For example, below we will change both the output directory and the output format of the resulting files.

```Bash
segul extract --dir alignments/ --input-format nexus --re="^Mus" output Mus-alignment/ --output-format Fasta
```

## Filtering alignments

`segul` provides multiple ways to filter alignments:

1. Based on alignment length
2. Based on the data matrix completeness.
3. Based on parsimony informative sites.

In its simplest form, the command is structure as below:

```Bash
segul <input-option> [alignment-path] <filtering-options>
```

> IMPORTANT NOTES: `segul` will never overwrite your original datasets. When filtering alignments, it will copy the alignments that match the filtering parameters to its output directory. You can also concat the resulting alignments instead. It is particularly useful to quickly estimate a species tree based on the filtered alignments.

### Filtering based on the minimal alignment length

To filter based on the alignment length (also refer to number of sites), we will use the `--len` option. For example, our directory below contains three alignments:

```Bash
alignments/
â”œâ”€â”€ locus_1.nexus
â”œâ”€â”€ locus_2.nexus
â””â”€â”€ locus_3.nexus
```

We would like to filter an alignment with minimal alignment length 500 bp. `segul` command using `--dir` input as below:

```Bash
segul --dir alignments/ --input-format nexus --len 500
```

We can achieve the same result using `--input` or `-i` option:

```Bash
segul --input alignments/*.nexus --len 500
```

### Filtering based on data matrix completeness

This filtering method is based on the percentage of minimal taxa present in your alignments, similar to `phyluce_align_get_only_loci_with_min_taxa` in [Phyluce pipeline](https://phyluce.readthedocs.io/en/latest/tutorials/tutorial-1.html#final-data-matrices). For example, given a collection of alignments with 100 taxa with 90 percent completeness, `segul` will filter alignments with at least containing 90 taxa.

To filter alignments based on data matrix completeness, we use `--percent` option with input in decimals. For example, here we will filter alignments in `alignments` directory with 80 percent data matrix completeness:

```Bash
segul --dir alignments/ --input-format nexus --percent 0.8
```

You can also write the percentage as `.8`, it will parse the same way as you write it with 0.

Sometimes, we want to try building phylogenetic trees with different data matrix completeness. Using `segul`, we can filter with multiple data matrix completeness values in a single commands. For this task, we use `npercent` option:

```Bash
segul --dir segul --dir alignments/ --input-format nexus --npercent .9 .8 .75
```

`segul` will create an output directory for each data matrix completeness value. The output directory names will be suffixed with the percentage values. For the example above, it will create three directories named `alignments_90p`, `alignments_80p`, and `alignments_75p`.

### Filtering based on parsimony informative sites

`segul` provides two ways in filtering based on parsimony informative sites. First, by inputing the minimal number of parsimony informative sites using the `--pinf` option. It will filter alignments that at least contains the input number of parsimony informative sites.

For example, in the command below, we will filter alignments in `alignments` directory with minimal contain 50 parsimony informative sites:

```Bash
segul filter --dir alignments/ --input-format nexus --pinf 50
```

Second, we can filter based on the percentage of minimal parsimony informative sites. This value is counted based on the highest number of parsimony informative sites in your alignments. It is similar to computing data matrix completeness, but based on the number of parsimony informative sites.

For example, the highest number of parsinomy informative sites in the input alignments is 100. We will filter with minimal 80 percent of the highest parsimony informative value. Using `--percent-inf .8` option, `segul` will filter an alignments with at least containing 80 parsimony informative sites:

```Bash
segul filter --dir alignments/ --input-format nexus --percent-inf .8
```

### Specifying output directories

By default, `segul` output directories for filtering are based on the input directory names suffixed with filtering values. The format as below:

```Text
[input-dir](underscore)[filtering-values]
```

For example, if your input directory is named `alignments/` and the filtering option is minimal alignment length of 500 bp, the output directory will be `alignments_500bp`.

You can change the prefix names using `--output` or `-o` option.

### Concatenating the results

By defaults, `segul` copy the alignments that match the filtering option. Instead of copying the files, you can concat them by using `--concat` flag. All of the options available for the `concat` subcommand are also available for this task ([see above](#concatenating-alignments)). You are, however, are required to specify the `--output` or `-o` name and the partition format `--part` or `-p` option. The output name will be used as the name of the output directory, the name of the concatenated alignment, and the prefix name for the partition setting file (if set for `raxml` or `nexus`).

For example, in the command below, we will filter based on alignment length of 500 bp and will concat the result:

```Bash
segul --dir alignments/ --input-format nexus --len 500 --concat --part raxml -output concat_alignment
```

## Finding unique IDs in alignments

Often, we need to know what are the taxa we in our dataset. The simplest command would be:

```bash
segul id -i [input-path]
```

`segul` will generate the ID in a text file containing a list of the ID in alphabetical order. The file will be saved as `id.txt` in your current working directory:

```Text
sequence_1
sequence_2
sequence_3
```

To change the file name, pass the name (without the extension) to the `--output` or `-o` option. For example:

```Bash
segul id -i alignments/ -o my_alignment_id
```

## Mapping sample distribution in a collection of alignments

To map the distribution of your samples across your dataset, you only need to pass `--map` flag in the finding unique IDs command:

```bash
segul id -i [input-path] --map
```

It will generate two files. The first file is the list of unique IDs in your dataset (named default to id.txt). This file is similar to generating unique IDs. The second file is a csv file (named default to id_map.txt) containing the distrubution of your samples presented in TRUE/FALSE values across your alignments. The content of the file will look like as below:

| Alignments | sequence_1 | sequence_2 | sequence_3 |
| ---------- | ---------- | ---------- | ---------- |
| locus_1    | TRUE       | FALSE      | TRUE       |
| locus_2    | TRUE       | TRUE       | TRUE       |
| locus_3    | FALSE      | FALSE      | TRUE       |

## Splitting concatenated alignments by partition

To split an alignment, you need two input files: the concatenated alignment and the partition file. If you skip inputing partition file, the app will assume the partition is embedded in the alignment file.

The command is as below:

```Bash
segul split --input <path-to-alignment> --input-partition <path-to-partition-file>
```

For example:

```Bash
segul split --input concat-alignment.nexus --input-partition concat-alignment-partition.nex
```

### Supported partition format for splitting alignment

`segul` support two kind of RaXML format:

With datatype written in the partitions:

```Text
DNA, locus_1 = 1-5
DNA, locus_2 = 6-10
```

Or without the datatype:

```Text
locus_1 = 1-5
locus_2 = 6-10
```

For nexus partition:

```Text
#Nexus
begin set
  charset locus_1 = 1-5;
  charset locus_2 = 6-10;
end;
```

`segul` partition parser is robust when dealing with extra spaces, but forbid the use of spaces for the locus names.

This nexus format will work:

```Text
#Nexus
begin set
  charset locus_1 = 1 - 5 ;
  charset locus_2 = 6 - 10 ;
end;
```

But this will not work:

```Text
#Nexus
begin set
  charset 'locus 1' = 1-5;
  charset 'locus 2' = 6-10;
end;
```

Some other programs write locus names with file extension (contain dot). In that case, `segul` will replace it with underscore to avoid name conflict when changing file extension.

For example:

```Text
#Nexus
begin set
  charset locus_1.nex = 1 - 5 ;
  charset locus_2.nex = 6 - 10 ;
end;
```

In this case above, `segul` will write the output file as locus_1_nex.nex and locus_2_nex.nex (if save to nexus).

## Batch renaming sequence IDs

`segul` provide an easy way to rename sequence IDs across all your alignments. To use this function, segul require a list of the original IDs and the the name IDs that it needs to change to. The input IDs can be written in a tabulated format as a comma-delimited file (`.csv`) or a tab-delimited file (`.tsv`).

The table must contain two columns with a header for each column. Both the header names and the IDs should contain no space. The name of the header does not matter. You can name it whichever makes sense for you. It is only there for record keeping. The order of the column, however, is important. The first column should be the original IDs and the second column should the new IDs. For example:

| Original_names        | New_names                |
| --------------------- | ------------------------ |
| Genus_species1_random | Genus_species1_voucherID |
| Genus_species2_random | Genus_species2_voucherID |

If you need to rename a lot of sequence ID, you can simplify the process by generating the unique IDs first ([see above](#finding-unique-ids-in-alignments)). Then, copy the IDs that you would like to rename in the `csv` or `tsv` input. To input this file, we will use the `--name` or `-n` option. `segul` will infer the file format based on its extension. The full command is structure as below:

```Bash
segul rename -d [alignment-dir] -f [sequence-format-keyword] -n [file-path-to-IDs]
```

For example:

```Bash
segul rename -d alignments/ -f nexus -n new_names.csv
```

You can use `--dry-run` option to check if the input names are parsed correctly:

```Bash
segul rename -d alignments/ -f nexus -n new_names.csv --dry-run
```

By default, `segul` will write the result in `SEGUL-Rename` directory. To change the output directory, use the `--output` or `-o` option.

```Bash
segul rename -d alignments/ --input-format nexus --name new_names.csv --output new_seq_names/
```

Like other functions, `segul` will not overwrite your original files. The output files will be default to nexus. To change the output format use the `--output-format` or `-F` function:

```Bash
segul rename -d alignments/ --input-format nexus --name new_names.csv --output new_seq_names/ --output-format fasta
```

## Translating DNA sequences

To translate dna alignment to amino acid:

```Bash
segul translate -d [path-to-alignment-files] -f [sequence-format-keyword]
```

By default, the app will use the standard code table (NCBI Table 1). To set the translation table, use the `--table` option. For example, to translate dna sequences using NCBI Table 2 (vertebrate MtDNA):

```Bash
segul translate -d loci/ -f fasta --table 2
```

You can also set the reading frame using the `--rf` option:

```Bash
segul translate -d loci/ -f fasta --table 2 --rf 2
```

To show all the table options, use the `--show-tables` flag:

```Bash
segul translate --show-tables
```

Supported [NCBI Genetic Code Tables](https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi#top):

| Table No | Genetic Code                                                                                 |
| -------- | -------------------------------------------------------------------------------------------- |
| 1        | The Standard Code                                                                            |
| 2        | The Vertebrate Mitochondrial Code                                                            |
| 3        | The Yeast Mitochondrial Code                                                                 |
| 4        | The Mold, Protozoan, and Coelenterate Mitochondrial Code and the Mycoplasma/Spiroplasma Code |
| 5        | The Invertebrate Mitochondrial Code                                                          |
| 6        | The Ciliate, Dasycladacean and Hexamita Nuclear Code                                         |
| 9        | The Echinoderm and Flatworm Mitochondrial Code                                               |
| 10       | The Euplotid Nuclear Code                                                                    |
| 11       | The Bacterial, Archaeal and Plant Plastid Code                                               |
| 12       | The Alternative Yeast Nuclear Code                                                           |
| 13       | The Ascidian Mitochondrial Code                                                              |
| 14       | The Alternative Flatworm Mitochondrial Code                                                  |
| 16       | Chlorophycean Mitochondrial Code                                                             |
| 21       | Trematode Mitochondrial Code                                                                 |
| 22       | Scenedesmus obliquus Mitochondrial Code                                                      |
| 23       | Thraustochytrium Mitochondrial Code                                                          |
| 24       | Rhabdopleuridae Mitochondrial Code                                                           |
| 25       | Candidate Division SR1 and Gracilibacteria Code                                              |
| 26       | Pachysolen tannophilus Nuclear Code                                                          |
| 29       | Mesodinium Nuclear Code                                                                      |
| 30       | Peritrich Nuclear Code                                                                       |
| 33       | Cephalodiscidae Mitochondrial UAA-Tyr Code                                                   |

## Log output

With the exception of the spinning emoji and the program progress messages, all the terminal output is written to `segul.log` file saved in the current working directory. In the log output, however, it is also include the time and the log status. For example, our terminal output is as folow:

```Text
=========================================================
SEGUL v0.12.0
An alignment tool for phylogenomics
---------------------------------------------------------
Input dir         : oliveros_et_al2019/
File counts       : 4,060
Input format      : Nexus
Data type         : DNA
Task              : Sequence format conversion

ðŸŒ˜ Finished converting sequence format!

Output
Output dir        : oliveros_et_al_2019_fasta
Output format     : Fasta Sequential
Log file          : segul.log

Execution time    : 1.762814458s
```

The log file is as follow:

```Text
2022-02-14 17:41:02 -06:00 - INFO - =========================================================
2022-02-14 17:41:02 -06:00 - INFO - SEGUL v0.12.0
2022-02-14 17:41:02 -06:00 - INFO - An alignment tool for phylogenomics
2022-02-14 17:41:02 -06:00 - INFO - ---------------------------------------------------------
2022-02-14 17:41:02 -06:00 - INFO - Input dir         : oliveros_et_al2019/
2022-02-14 17:41:02 -06:00 - INFO - File counts       : 4,060
2022-02-14 17:41:02 -06:00 - INFO - Input format      : Nexus
2022-02-14 17:41:02 -06:00 - INFO - Data type         : DNA
2022-02-14 17:41:02 -06:00 - INFO - Task              : Sequence format conversion

2022-02-14 17:41:04 -06:00 - INFO - Output
2022-02-14 17:41:04 -06:00 - INFO - Output dir        : oliveros_et_al_2019_fasta
2022-02-14 17:41:04 -06:00 - INFO - Output format     : Fasta Sequential
2022-02-14 17:41:04 -06:00 - INFO - Log file          : segul.log
2022-02-14 17:41:04 -06:00 - INFO - Execution time    : 1.762814458s
```

`segul` does not overwrite the log file. Each time you run the app, if the log file exists in the same working directory, it will append the log output into the same file. If you would like to keep a different log file for a different task, you either move the file to a different folder or rename it.
